linter errors when multiple handlers are installed simultaneously

code such as 
```python
with handler({get_proposal: auto_uniform_proposal, norm: make_l2_norm(10)}):...
```
throws a linter error:
```bash
tests/test_ops/test_loss.py:248: error: Argument 1 to "handler" has incompatible type "dict[object, Callable[..., Any]]"; expected "Mapping[Operation[[VarArg(Any), KwArg(Any)], Never], Callable[..., Any]]"  [arg-type]
```

using mypy version 1.13.0

--- a/effectful/ops/semantics.py
+++ b/effectful/ops/semantics.py
@@ -15,10 +15,8 @@ T = TypeVar("T")
 V = TypeVar("V")
 
 
-@defop  # type: ignore
-def apply(
-    intp: Interpretation[S, T], op: Operation[P, S], *args: P.args, **kwargs: P.kwargs
-) -> T:
+@defop
+def apply(intp: Interpretation, op: Operation, *args, **kwargs) -> Any:
     """Apply ``op`` to ``args``, ``kwargs`` in interpretation ``intp``.
 
     Handling :func:`apply` changes the evaluation strategy of terms.
@@ -50,7 +48,7 @@ def apply(
     elif apply in intp:
         return intp[apply](intp, op, *args, **kwargs)
     else:
-        return op.__default_rule__(*args, **kwargs)  # type: ignore
+        return op.__default_rule__(*args, **kwargs)
 
 
 @defop  # type: ignore
@@ -92,9 +90,7 @@ def fwd(*args, **kwargs) -> Any:
     raise RuntimeError("fwd should only be called in the context of a handler")
 
 
-def coproduct(
-    intp: Interpretation[S, T], intp2: Interpretation[S, T]
-) -> Interpretation[S, T]:
+def coproduct(intp: Interpretation, intp2: Interpretation) -> Interpretation:
     """The coproduct of two interpretations handles any effect that is handled
     by either. If both interpretations handle an effect, ``intp2`` takes
     precedence.
@@ -150,7 +146,7 @@ def coproduct(
         if op is fwd or op is _get_args:
             res[op] = i2  # fast path for special cases, should be equivalent if removed
         else:
-            i1 = intp.get(op, op.__default_rule__)  # type: ignore
+            i1 = intp.get(op, op.__default_rule__)
 
             # calling fwd in the right handler should dispatch to the left handler
             res[op] = _set_prompt(fwd, _restore_args(_save_args(i1)), _save_args(i2))
@@ -158,9 +154,7 @@ def coproduct(
     return res
 
 
-def product(
-    intp: Interpretation[S, T], intp2: Interpretation[S, T]
-) -> Interpretation[S, T]:
+def product(intp: Interpretation, intp2: Interpretation) -> Interpretation:
     """The product of two interpretations handles any effect that is handled by
     ``intp2``. Handlers in ``intp2`` may override handlers in ``intp``, but
     those changes are not visible to the handlers in ``intp``. In this way,
@@ -206,7 +200,7 @@ def product(
 
 
 @contextlib.contextmanager
-def runner(intp: Interpretation[S, T]):
+def runner(intp: Interpretation):
     """Install an interpretation by taking a product with the current
     interpretation.
 
@@ -222,7 +216,7 @@ def runner(intp: Interpretation[S, T]):
 
 
 @contextlib.contextmanager
-def handler(intp: Interpretation[S, T]):
+def handler(intp: Interpretation):
     """Install an interpretation by taking a coproduct with the current
     interpretation.
 
@@ -233,7 +227,7 @@ def handler(intp: Interpretation[S, T]):
         yield intp
 
 
-def evaluate(expr: Expr[T], *, intp: Optional[Interpretation[S, T]] = None) -> Expr[T]:
+def evaluate(expr: Expr[T], *, intp: Optional[Interpretation] = None) -> Expr[T]:
     """Evaluate expression ``expr`` using interpretation ``intp``. If no
     interpretation is provided, uses the current interpretation.
 
@@ -261,7 +255,7 @@ def evaluate(expr: Expr[T], *, intp: Optional[Interpretation[S, T]] = None) -> E
         (args, kwargs) = tree.map_structure(
             functools.partial(evaluate, intp=intp), (expr.args, expr.kwargs)
         )
-        return apply.__default_rule__(intp, expr.op, *args, **kwargs)  # type: ignore
+        return apply.__default_rule__(intp, expr.op, *args, **kwargs)
     elif tree.is_nested(expr):
         return tree.map_structure(functools.partial(evaluate, intp=intp), expr)
     else:

--- a/tests/test_ops_semantics.py
+++ b/tests/test_ops_semantics.py
@@ -6,7 +6,16 @@ from typing import TypeVar
 import pytest
 from typing_extensions import ParamSpec
 
-from effectful.ops.semantics import coproduct, evaluate, fvsof, fwd, handler, product
+from effectful.ops.semantics import (
+    apply,
+    coproduct,
+    evaluate,
+    fvsof,
+    fwd,
+    handler,
+    product,
+    runner,
+)
 from effectful.ops.syntax import ObjectInterpretation, defop, implements
 from effectful.ops.types import Interpretation, Operation
 
@@ -552,3 +561,44 @@ def test_ctxof():
     assert fvsof(Nested([x()], y())) >= {x, y}
     assert fvsof(Nested([x()], [y()])) >= {x, y}
     assert fvsof(Nested((x(), y()))) >= {x, y}
+
+
+def test_handler_typing() -> None:
+    """This test is for the type checker; it doesn't do anything interesting
+    when run.
+
+    """
+
+    @defop
+    def f(x: int) -> int:
+        raise NotImplementedError
+
+    @defop
+    def g(x: str, y: bool) -> str:
+        return "test" if y else x
+
+    # Note: this annotation is required. Without annotation, mypy joins the two
+    # operator types to `object`.
+    i: Interpretation = {f: lambda x: x + 1, g: lambda x, y: x + str(y)}
+
+    handler(i)
+    runner(i)
+    product(i, i)
+    coproduct(i, i)
+    apply(i, f, 1)
+    evaluate(0, intp=i)
+
+    # include tests with inlined interpretation, because mypy might do inference
+    # differently
+    handler({f: lambda x: x + 1, g: lambda x, y: x + str(y)})
+    runner({f: lambda x: x + 1, g: lambda x, y: x + str(y)})
+    product(
+        {f: lambda x: x + 1, g: lambda x, y: x + str(y)},
+        {f: lambda x: x + 1, g: lambda x, y: x + str(y)},
+    )
+    coproduct(
+        {f: lambda x: x + 1, g: lambda x, y: x + str(y)},
+        {f: lambda x: x + 1, g: lambda x, y: x + str(y)},
+    )
+    apply({f: lambda x: x + 1, g: lambda x, y: x + str(y)}, f, 1)
+    evaluate(0, intp={f: lambda x: x + 1, g: lambda x, y: x + str(y)})

